### PR TITLE
URLBear: Use library to extract links

### LINKED ISSUE #1342: InvalidLinkBear: use library to extract links
As we hit more corner cases for url extraction, our regex approach will be a limiting factor, and we should switch to using an external library. We would need to import our regression tests into the external library.

https://pypi.python.org/pypi/urlextract might be useful.

Or we may find a 'recursive' web scraper which handles file types of than .html.
I.e. there maybe a function in https://scrapy.org which handles plain text documents.

### PR DESCRIPTION
**Pull request by CLiu13**
This replaces the use of regex for extracting
links with the use of the URLExtract library.

Closes https://github.com/coala/coala-bears/issues/1342

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [ ] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [ ] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `corobo mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

File: bear-requirements.txt [hunk 1]
```diff
@@ -36,6 +36,7 @@ rstcheck~=3.1
 safety~=1.8.2
 scspell3k~=2.0
 sqlparse~=0.2.4
+git+https://github.com/lipoja/URLExtract.git
```
Comment (refeed) [inline bear-requirements.txt]: `URLExtract` has a pip package on https://pypi.org/project/urlextract/ . Could you explain why it's necessary to use the `git+https://` form?

File: tests/general/URLBearTest.py [hunk 2]
```diff
@@ -37,6 +37,16 @@ def test_detect_url_result(self):
                          [3, 'http://www.google.com/404',
                           LINK_CONTEXT.no_context])
 
+    def test_detect_pip_vcs_url_result(self):
+        valid_file = """
+        git+http://www.github.com/foo
```
Comment (refeed) [inline tests/general/URLBearTest.py]: Just a nitpicky comment. `git+http` pip vcs url usually has a `.git` suffix in it ;)

File: bear-requirements.txt (hunk 1 above)
Comment (CLiu13) [reply bear-requirements.txt]: The version of URLExtract on pypi has not been updated in several months, and I need the latest version with the latest changes in order to implement this new feature.

File: tests/general/URLBearTest.py [hunk 3]
```diff
@@ -53,6 +63,42 @@ def test_precentage_encoded_url(self):
...
+
+        result = get_results(self.uut, valid_file)
+        self.assertEqual(result[0].contents,
+                         [2, 'http://google.com/trailing',
+                          LINK_CONTEXT.no_context])
+
+    def test_detect_example_url_result(self):
+        invalid_file = """
+        http://example.com
+        """.splitlines()
```
Comment (li-boxuan) [inline tests/general/URLBearTest.py]: Why is `example.com` an invalid url?

Comment (CLiu13) [reply tests/general/URLBearTest.py]: The existing code for URLBear allows the user to input regex that specifies certain links to ignore, and [the run() function of URLBear specifies `example.com` as a link to ignore](https://github.com/coala/coala-bears/blob/2888639aa2427ae77ab74486e8750770a1e219c1/bears/general/URLBear.py#L135), hence why I marked `example.com` as invalid.

File: bear-requirements.txt (hunk 1 above)
Comment (CLiu13) [reply bear-requirements.txt]: Sure thing! Just sent an email to the creator.

**Comment by jayvdb** [comment]
@gitmate-bot rebase

**Comment by jayvdb** [comment]
@gitmate-bot rebase

**Comment by jayvdb** [comment]
Tests are failing.

**Comment by CLiu13** [comment]
> Tests are failing.

Failures are due to the use of URLExtract in `requirements.txt` since [it's listed directly to its source on GitHub](https://github.com/coala/coala-bears/pull/2822/files#diff-0f48fb9283f606dbaa9ae7328bd9c4b1R39), rather than to its PyPi release (which is outdated and missing important changes). However, [the owner has commented](https://github.com/lipoja/URLExtract/issues/29#issuecomment-457350615) that they will do something about releasing a new version soon.

File: bear-requirements.txt (hunk 1 above)
Comment (CLiu13) [reply bear-requirements.txt]: The creator released a [new version on PyPi](https://pypi.org/project/urlextract/0.9/) yesterday, and I have updated `bear-requirements.txt` accordingly.
